pandas/src/inference.pyx

```diff
@@ -85,6 +85,129 @@ try:
+        bint snumeric        # seen_numeric
+        bint sdatetime       # seen_datetime
+        bint coerce_numeric  # coerce data to numeric
+        bint stimedelta      # seen_timedelta
+        bint sdatetimetz     # seen_datetimetz
+
+    def __cinit__(self, bint coerce_numeric=0):
+        """
+        Initialize a Seen instance.
+
+        Parameters
+        ----------
+        coerce_numeric : bint, default 0
+            Whether or not to force conversion to a numeric data type if
+            initial methods to convert to numeric fail.
+        """
+        self.sint = 0
+        self.sbool = 0
+        self.snull = 0
+        self.suint = 0
+        self.ssint = 0
+        self.sfloat = 0
+        self.sobject = 0
+        self.scomplex = 0
+        self.snumeric = 0
+        self.sdatetime = 0
+        self.stimedelta = 0
+        self.sdatetimetz = 0
+        self.coerce_numeric = coerce_numeric
+
+    cdef inline bint check_uint64_nan(self) except -1:
+        """
+        Check whether we have encountered uint64 when handling a NaN element.
+
+        If uint64 has been encountered, we cannot safely cast to float64 due
+        to truncation problems (this would occur if we return a numeric array
+        containing a NaN element).
+
+        Returns
+        -------
+        return_values : bool
+            Whether or not we should return the original input array to avoid
+            data truncation.
+
+        Raises
+        ------
+        ValueError : uint64 elements were detected, and an element is being
+                     coerced to NaN, which does not exist in uint64.
+        """
+        if self.snull and self.suint:
+            if not self.coerce_numeric:
+                return True
+            else:
+                raise ValueError("uint64 array detected, and such an "
+                                 "array cannot contain NaN.")
+
+        return False
+
+    cdef inline bint check_uint64_int64_conflict(self) except -1:
+        """
+        Check whether we have encountered both int64 and uint64 elements.
+
+        If both have been encountered, we cannot safely cast to an integer
+        dtype since none is large enough to hold both types of elements.
+
+        Returns
+        -------
+        return_values : bool
+            Whether or not we should return the original input array to avoid
+            data truncation.
+
+        Raises
+        ------
+        ValueError : negative and uint64 elements were detected, and returning
+                     a numeric array is being forced.
+        """
+        if self.ssint and self.suint:
+            if not self.coerce_numeric:
+                return True
+            else:
+                raise ValueError("uint64 and negative values detected. "
+                                 "Cannot safely return a numeric array "
+                                 "without truncating data.")
+
+        return False
+
+    cdef inline saw_null(self):
```

Comment:
I *think*, and maybe you have a better refence. that it doesn't matter if you cdef a method in a cdef class, by-definiiion they are *always* cdef. e.g. in ``Timestamp`` we mark these ``def`` actually. 

I *also* am not sure of the ``inline`` when in a cdef class (IOW it may or may not make any difference).